spec/abci++/abci++_basic_concepts_002_draft.md

```diff
@@ -0,0 +1,372 @@
+successive `FinalizeBlock` calls. This is critical, because the
+responses are included in the header of the next block, either via a Merkle root
+or directly, so all nodes must agree on exactly what they are.
+
+For this reason, it is recommended that applications not be exposed to any
+external user or process except via the ABCI connections to a consensus engine
+like Tendermint Core. The application must only change its state based on input
+from block execution (`FinalizeBlock` calls), and not through
+any other kind of request. This is the only way to ensure all nodes see the same
+transactions and compute the same results.
+
+Some Applications may choose to execute the blocks that are about to be proposed
+(via `PrepareProposal`), or those that the application is asked to validate
+(via `Processproposal`). However the state changes caused by processing those
+proposed blocks must never replace the previous state until `FinalizeBlock` confirms
+the block decided.
+
+Additionally, vote extensions or the validation thereof (via `ExtendVote` or
+`VerifyVoteExtension`) must _never_ have side effects on the current state.
+They can only be used when their data is included in a block.
+
+If there is some non-determinism in the state machine, consensus will eventually
+fail as nodes disagree over the correct values for the block header. The
+non-determinism must be fixed and the nodes restarted.
+
+Sources of non-determinism in applications may include:
+
+* Hardware failures
+    * Cosmic rays, overheating, etc.
+* Node-dependent state
+    * Random numbers
+    * Time
+* Underspecification
+    * Library version changes
+    * Race conditions
+    * Floating point numbers
+    * JSON serialization
+    * Iterating through hash-tables/maps/dictionaries
+* External Sources
+    * Filesystem
+    * Network calls (eg. some external REST API service)
+
+See [#56](https://github.com/tendermint/abci/issues/56) for original discussion.
+
+Note that some methods (`Query, CheckTx, FinalizeBlock`) return
+explicitly non-deterministic data in the form of `Info` and `Log` fields. The `Log` is
+intended for the literal output from the application's logger, while the
+`Info` is any additional info that should be returned. These are the only fields
+that are not included in block header computations, so we don't need agreement
+on them. All other fields in the `Response*` must be strictly deterministic.
+
+## Block Execution
+
+The first time a new blockchain is started, Tendermint calls
+`InitChain`. From then on, method `FinalizeBlock` is executed at the end of each
+block, resulting in an updated application state.
+During consensus execution of a block height, before method `FinalizeBlock` is
+called, methods `PrepareProposal`, `ProcessProposal`, `ExtendVote`, and
+`VerifyVoteExtension` may be called a number of times.
+See [Application Requirements](abci++_app_requirements_002_draft.md) for
+details on the possible call sequences of these methods.
```

Comment:
Shouldn't that point to "expected behaviors"?

Reply:
You're right! Fixing it...